nexus/src/internal_api/http_entrypoints.rs

```diff
@@ -280,3 +285,84 @@ async fn cpapi_metrics_collect(
         .instrument_dropshot_handler(&request_context, handler)
         .await
 }
+
+#[derive(Deserialize, JsonSchema)]
+struct AllPath {
+    path: String,
+}
+
+/// Endpoint used by Sled Agents to download cached artifacts.
+#[endpoint {
+    method = GET,
+    path = "/artifacts/{path}",
+}]
+async fn cpapi_artifact_download(
+    request_context: Arc<RequestContext<Arc<ServerContext>>>,
+    path: Path<AllPath>,
+) -> Result<Response<Body>, HttpError> {
+    let context = request_context.context();
+    let nexus = &context.nexus;
+    let mut entry = PathBuf::from(BASE_ARTIFACT_DIR);
+
+    // TODO: Most of the below code is ready to accept a multi-component path,
+    // such as in:
+    // https://github.com/oxidecomputer/dropshot/blob/78be3deda556a9339ea09f3a9961fd91389f8757/dropshot/examples/file_server.rs#L86-L89
+    //
+    // However, openapi does *not* like that currently, so we limit the endpoint
+    // to only accepting single-component paths.
+    let path = vec![path.into_inner().path];
+
+    let handler = async {
+        for component in &path {
```

Comment:
if we can find a common logic maybe this can be shared with `find_file` in `console_api`

Reply:
We're doing something a little quirky here - basically, lazily fetching files that might not (yet) exist locally. Dunno if that' easy to dedup or not, but I'd be interested in finding commonalities between the two paths.

Reply:
Yeah, I see. We bail if we get to a path segment that doesn't exist, while you can keep going through and building up the path regardless. Hard to see how to share that logic in a way that doesn't make both use cases harder to understand.